### ISSUE TITLE
Control Keyboard from application

### ISSUE THREAD

**Issue by mirij100** [florisboard/florisboard#951]
Hello.

Is it possible to control this keyboard from application (on non rooted device)?
I need the following:
- switching keyboard layout (for example between Czech, Polish, Hungary etc.)
- prevent user from changing keyboard layout by long press Space key
- prevent keyboard floating mode
- hide (don't show) the upper line with icons
- set the y-position of the keyboard - I need to show my custom text/button under the keyboard 

Regards, Jiří Mikeska

**Comment by patrickgold** [comment]
FlorisBoard currently does not support the `privateImeOptions` feature, as this is mainly intended for usage between an app and a keyboard which are from the same developer/company/project etc. Using this feature is of course possible from all apps, but I am a bit skeptical from a security standpoint that any unknown app can semi-control the keyboard in such a way.

Some things (like setting the y position) may be considerable to adapt the keyboard to the app layout, but preventing the user from changing the keyboard via space bar is a nono (at any point the user of the keyboard must be able to switch to another keyboard, as FlorisBoard supports the `switchToNextIme` attribute and enforces this).

Do you have any example of another (open-source) keyboard which supports this feature and link it to me? Maybe I try to see what keyboards offer publicly to use this as a guideline for an implementation within FlorisBoard.

**Comment by mirij100** [comment]
> Do you have any example of another (open-source) keyboard which supports this feature and link it to me?

On the contrary, I am looking for the keyboard which support software control these things. I am developing kiosk application for Android and need these features. Throught privateImeOptions is possible to control only some small things on standard preinstalled Google Gboard keyboard.

**Comment by patrickgold** [comment]
Ok I see. Still I am a bit skeptical when it comes to such lengths for a app controlling the keyboard. Will place it on my maybe list for now and reevaluate the options for implementing such feature at a later stage.

**Comment by kj7rrv** [comment]
Could this be an option under Advanced settings so only people who really need it turn it on?

**Comment by patrickgold** [comment]
> Could this be an option under Advanced settings so only people who really need it turn it on?

Yes could be but the question is how many users will actually turn such feature on? Maybe if I add a control toggle for each application (so only some whitelisted apps can have access is more useful, don't know though).

**Comment by mirij100** [comment]
For me would be ideal, no control toggle, no whitelist. Simply install package, for example by adb.exe, and run my application (no additional configuration action from user). And then control these features from my application throught `<EditText android:imeOptions="xxx" />` or `<EditText android:privateImeOptions="xxx" />`.

**Comment by kj7rrv** [comment]
Okay, that makes sense. It does sound better to allow it per-app.

On Thursday, June 03, 2021 08:56 PDT, Patrick Goldinger ***@***.***> wrote:
 
 
Could this be an option under Advanced settings so only people who really need it turn it on?
Yes could be but the question is how many users will actually turn such feature on? Maybe if I add a control toggle for each application (so only some whitelisted apps can have access is more useful, don't know though).
